File: packages/control-plane/loop_control_plane/tool_contracts.py
from __future__ import annotations

import asyncio
import json
from datetime import UTC, datetime
from hashlib import sha256
from typing import Any, Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field

from loop_control_plane._app_agents import AgentRecord
from loop_control_plane.workspaces import WorkspaceError

ToolSideEffectLevel = Literal[
    "unknown",
    "read",
    "write",
    "money_movement",
    "external_message",
]
SandboxStatus = Literal["mock", "sandbox", "disabled"]
LiveStatus = Literal["disabled", "review_required", "approved", "blocked"]


class ToolContractUpsert(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=160)
    description: str = Field(default="", max_length=1200)
    side_effect_level: ToolSideEffectLevel = "unknown"
    pii_access: bool = False
    money_movement: bool = False
    rate_limits: dict[str, Any] = Field(default_factory=dict)
    budget_limits: dict[str, Any] = Field(default_factory=dict)
    sandbox_status: SandboxStatus = "sandbox"
    owner_user_id: str = Field(default="", max_length=160)
    approval_policy_id: str = Field(default="", max_length=160)
    failure_behavior: str = Field(default="", max_length=1200)
    compensation_behavior: str = Field(default="", max_length=1200)


class ToolContractRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    workspace_id: UUID
    agent_id: UUID
    tool_id: str
    name: str
    description: str
    side_effect_level: ToolSideEffectLevel
    pii_access: bool
    money_movement: bool
    rate_limits: dict[str, Any]
    budget_limits: dict[str, Any]
    sandbox_status: SandboxStatus
    live_status: LiveStatus
    owner_user_id: str
    approval_policy_id: str
    failure_behavior: str
    compensation_behavior: str
    content_hash: str
    approval_invalidated_at: datetime | None
    created_at: datetime
    updated_at: datetime


def _content_hash(body: ToolContractUpsert) -> str:
    payload = {
        "name": body.name,
        "description": body.description,
        "side_effect_level": body.side_effect_level,
        "pii_access": body.pii_access,
        "money_movement": body.money_movement,
        "rate_limits": body.rate_limits,
        "budget_limits": body.budget_limits,
        "sandbox_status": body.sandbox_status,
        "owner_user_id": body.owner_user_id,
        "approval_policy_id": body.approval_policy_id,
        "failure_behavior": body.failure_behavior,
        "compensation_behavior": body.compensation_behavior,
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":"), default=str)
    return sha256(encoded.encode("utf-8")).hexdigest()


def _live_status_for(body: ToolContractUpsert) -> LiveStatus:
    if body.side_effect_level == "unknown":
        return "review_required"
    if body.money_movement and not body.budget_limits:
        return "blocked"
    if body.side_effect_level != "read" and not body.failure_behavior:
        return "review_required"
    return "disabled"


def _promotion_blocker(record: ToolContractRecord) -> str | None:
    if record.side_effect_level == "unknown":
        return "side-effect classification is required"
    if record.money_movement and not record.budget_limits:
        return "money-moving tools require budget caps"
    if record.side_effect_level != "read" and not record.failure_behavior:
        return "mutating tools require failure behavior"
    if record.money_movement and not record.compensation_behavior:
        return "money-moving tools require compensation behavior"
    if not record.owner_user_id:
        return "tool owner is required"
    return None


def tool_contract_payload(record: ToolContractRecord) -> dict[str, Any]:
    return record.model_dump(mode="json")


class ToolContractRegistry:
    def __init__(self) -> None:
        self._contracts: dict[UUID, list[ToolContractRecord]] = {}
        self._lock = asyncio.Lock()

    async def list_for_agent(self, *, agent: AgentRecord) -> list[ToolContractRecord]:
        async with self._lock:
            return list(self._contracts.get(agent.id, []))

    async def upsert(
        self,
        *,
        agent: AgentRecord,
        tool_id: str,
        body: ToolContractUpsert,
    ) -> ToolContractRecord:
        async with self._lock:
            now = datetime.now(UTC)
            contracts = self._contracts.setdefault(agent.id, [])
            content_hash = _content_hash(body)
            existing_index = next(
                (index for index, contract in enumerate(contracts) if contract.tool_id == tool_id),
                None,
            )
            if existing_index is None:
                record = ToolContractRecord(
                    id=f"tc_{uuid4().hex[:12]}",
                    workspace_id=agent.workspace_id,
                    agent_id=agent.id,
                    tool_id=tool_id,
                    name=body.name,
                    description=body.description,
                    side_effect_level=body.side_effect_level,
                    pii_access=body.pii_access,
                    money_movement=body.money_movement,
                    rate_limits=body.rate_limits,
                    budget_limits=body.budget_limits,
                    sandbox_status=body.sandbox_status,
                    live_status=_live_status_for(body),
                    owner_user_id=body.owner_user_id,
                    approval_policy_id=body.approval_policy_id,
                    failure_behavior=body.failure_behavior,
                    compensation_behavior=body.compensation_behavior,
                    content_hash=content_hash,
                    approval_invalidated_at=None,
                    created_at=now,
                    updated_at=now,
                )
                contracts.insert(0, record)
                return record

            existing = contracts[existing_index]
            approval_invalidated_at = existing.approval_invalidated_at
            live_status = _live_status_for(body)
            if existing.live_status == "approved" and existing.content_hash != content_hash:
                approval_invalidated_at = now
                live_status = "review_required"
            record = existing.model_copy(
                update={
                    "name": body.name,
                    "description": body.description,
                    "side_effect_level": body.side_effect_level,
                    "pii_access": body.pii_access,
                    "money_movement": body.money_movement,
                    "rate_limits": body.rate_limits,
                    "budget_limits": body.budget_limits,
                    "sandbox_status": body.sandbox_status,
                    "live_status": live_status,
                    "owner_user_id": body.owner_user_id,
                    "approval_policy_id": body.approval_policy_id,
                    "failure_behavior": body.failure_behavior,
                    "compensation_behavior": body.compensation_behavior,
                    "content_hash": content_hash,
                    "approval_invalidated_at": approval_invalidated_at,
                    "updated_at": now,
                }
            )
            contracts[existing_index] = record
            return record

    async def promote(self, *, agent: AgentRecord, tool_id: str) -> ToolContractRecord:
        async with self._lock:
            contracts = self._contracts.get(agent.id, [])
            for index, contract in enumerate(contracts):
                if contract.tool_id != tool_id:
                    continue
                blocker = _promotion_blocker(contract)
                if blocker is not None:
                    raise WorkspaceError(blocker)
                updated = contract.model_copy(
                    update={
                        "live_status": "approved",
                        "approval_invalidated_at": None,
                        "updated_at": datetime.now(UTC),
                    }
                )
                contracts[index] = updated
                return updated
        raise WorkspaceError(f"unknown tool contract: {tool_id}")

File: apps/studio/src/lib/tool-contracts.ts
import { cpJson, type UxWireupClientOptions } from "@/lib/ux-wireup";

export type ToolSideEffectLevel =
  | "unknown"
  | "read"
  | "write"
  | "money_movement"
  | "external_message";
export type ToolSandboxStatus = "mock" | "sandbox" | "disabled";
export type ToolLiveStatus =
  | "disabled"
  | "review_required"
  | "approved"
  | "blocked";

export interface ToolContract {
  id: string;
  workspace_id: string;
  agent_id: string;
  tool_id: string;
  name: string;
  description: string;
  side_effect_level: ToolSideEffectLevel;
  pii_access: boolean;
  money_movement: boolean;
  rate_limits: Record<string, unknown>;
  budget_limits: Record<string, unknown>;
  sandbox_status: ToolSandboxStatus;
  live_status: ToolLiveStatus;
  owner_user_id: string;
  approval_policy_id: string;
  failure_behavior: string;
  compensation_behavior: string;
  content_hash: string;
  approval_invalidated_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ToolContractsResponse {
  items: ToolContract[];
}

export type ToolContractInput = Pick<
  ToolContract,
  | "name"
  | "description"
  | "side_effect_level"
  | "pii_access"
  | "money_movement"
  | "rate_limits"
  | "budget_limits"
  | "sandbox_status"
  | "owner_user_id"
  | "approval_policy_id"
  | "failure_behavior"
  | "compensation_behavior"
>;

function localContract(agentId: string, toolId: string): ToolContract {
  const now = new Date(0).toISOString();
  const isRefund = /refund|payment|charge/i.test(toolId);
  return {
    id: `tc_local_${toolId}`,
    workspace_id: "local",
    agent_id: agentId,
    tool_id: toolId,
    name: toolId,
    description: isRefund
      ? "Create a refund only after policy and approval checks."
      : "Read-only lookup tool contract.",
    side_effect_level: isRefund ? "money_movement" : "read",
    pii_access: false,
    money_movement: isRefund,
    rate_limits: { per_minute: isRefund ? 20 : 600 },
    budget_limits: isRefund ? { max_per_call_cents: 50_000 } : {},
    sandbox_status: "sandbox",
    live_status: isRefund ? "review_required" : "approved",
    owner_user_id: "workspace-builder",
    approval_policy_id: isRefund ? "policy-money-movement" : "policy-read-only",
    failure_behavior: isRefund
      ? "Escalate instead of promising a refund."
      : "Answer with uncertainty if lookup fails.",
    compensation_behavior: isRefund
      ? "Void pending refund when downstream write fails."
      : "No compensation required for read-only calls.",
    content_hash: `local_hash_${toolId}`,
    approval_invalidated_at: null,
    created_at: now,
    updated_at: now,
  };
}

export function localToolContracts(
  agentId: string,
  toolIds: string[],
): ToolContract[] {
  return toolIds.map((toolId) => localContract(agentId, toolId));
}

export async function listToolContracts(
  agentId: string,
  opts: UxWireupClientOptions = {},
): Promise<ToolContractsResponse> {
  return cpJson<ToolContractsResponse>(
    `/agents/${encodeURIComponent(agentId)}/tool-contracts`,
    {
      ...opts,
      fallback: { items: [] },
    },
  );
}

export async function upsertToolContract(
  agentId: string,
  toolId: string,
  input: ToolContractInput,
  opts: UxWireupClientOptions = {},
): Promise<ToolContract> {
  return cpJson<ToolContract>(
    `/agents/${encodeURIComponent(agentId)}/tool-contracts/${encodeURIComponent(
      toolId,
    )}`,
    {
      ...opts,
      method: "PUT",
      body: input,
      fallback: {
        ...localContract(agentId, toolId),
        ...input,
        tool_id: toolId,
        updated_at: new Date(0).toISOString(),
      },
    },
  );
}

export async function promoteToolContract(
  agentId: string,
  toolId: string,
  opts: UxWireupClientOptions = {},
): Promise<ToolContract> {
  const fallback = localContract(agentId, toolId);
  return cpJson<ToolContract>(
    `/agents/${encodeURIComponent(agentId)}/tool-contracts/${encodeURIComponent(
      toolId,
    )}/promote`,
    {
      ...opts,
      method: "POST",
      fallback: {
        ...fallback,
        live_status: "approved",
        approval_invalidated_at: null,
      },
    },
  );
}
